File: tracking/tracking.php
<?php

// Dettagli del server SFTP
$hostname = 'ws.italagro.bindcommerce.biz';  
$port = 22;  
$username = 'wsitalagro';  
$password = 'Q0W80q8oeuKWztztdTd2QL5JphA7lWgP';  

// Percorsi dei file
$remote_file_path = '/home/wsitalagro/webapps/ws-italagro/orders/export_orders.csv';  // File remoto da leggere
$output_file_path = '/home/wsitalagro/webapps/ws-italagro/tracking/export_expedition.csv';  // File CSV da salvare

// Creazione della connessione SFTP
$connection = ssh2_connect($hostname, $port);

if (!$connection) {
    die('Impossibile connettersi al server SFTP.');
}

// Autenticazione con nome utente e password
if (!ssh2_auth_password($connection, $username, $password)) {
    die('Autenticazione fallita.');
}

// Apertura della connessione SFTP
$sftp = ssh2_sftp($connection);

// Verifica se il file remoto esiste
$remote_file = "ssh2.sftp://$sftp$remote_file_path";
if (!file_exists($remote_file)) {
    die("Il file remoto non esiste: $remote_file_path");
}

// Leggi il contenuto del file CSV remoto
$file_content = file_get_contents($remote_file);

// Carica il contenuto CSV in un array
$lines = explode("\n", $file_content);
$header = str_getcsv(array_shift($lines));  // Estrai l'header

// Trova gli indici delle colonne richieste 
$columns_to_select = ['Order Number', 'Tracking Number', 'Carrier', 'OrderDate'];
$selected_columns_indices = [];
foreach ($header as $index => $column_name) {
    if (in_array($column_name, $columns_to_select)) {
        $selected_columns_indices[] = $index;
    }
}

// Seleziona i dati delle colonne desiderate
$selected_data = [];
foreach ($lines as $line) {
    $data = str_getcsv($line);
    if ($data) {
        $selected_data[] = [
            $data[$selected_columns_indices[0]],
            $data[$selected_columns_indices[1]],
            $data[$selected_columns_indices[2]],
            $data[$selected_columns_indices[3]],  
        ];
    }
}

// Crea il contenuto del nuovo file CSV
$new_csv_content = implode(",", $columns_to_select) . "\n";
foreach ($selected_data as $row) {
    $new_csv_content .= implode(",", $row) . "\n";
}

// Salva il nuovo file CSV nella cartella remota
$remote_output_file = "ssh2.sftp://$sftp$output_file_path";
file_put_contents($remote_output_file, $new_csv_content);

echo "File CSV selezionato salvato con successo in: $output_file_path";

// Chiudi la connessione SFTP
ssh2_disconnect($connection);

?>

File: tracking/tracking.py
import paramiko
import pandas as pd
from io import StringIO
import schedule
import time

# Funzione per eseguire il salvataggio del CSV
def salva_csv():
    # Dettagli del server SFTP
    hostname = 'ws.italagro.bindcommerce.biz'  
    port = 22 
    username = 'wsitalagro'  
    password = 'Q0W80q8oeuKWztztdTd2QL5JphA7lWgP'  

    # Percorsi dei file
    remote_file_path = '/home/wsitalagro/webapps/ws-italagro/tracking/tracking_vg.csv'  # File remoto da leggere
    output_file_path = '/home/wsitalagro/webapps/ws-italagro/tracking/export_tracking_vg.csv'  # File CSV da salvare

    # Connessione al server SFTP
    try:
        transport = paramiko.Transport((hostname, port))
        transport.connect(username=username, password=password)
        print("Connessione al server SFTP in corso....")

        # Crea il client SFTP
        sftp = paramiko.SFTPClient.from_transport(transport)

        # Leggi il contenuto del file remoto
        print("Leggo il CSV dei Tracking...")
        with sftp.open(remote_file_path, 'r') as remote_file:
            file_content = remote_file.read().decode('utf-8')

        # Crea un DataFrame Pandas dal contenuto CSV
        df = pd.read_csv(StringIO(file_content))

        # Seleziona solo le colonne richieste 
        print("Genero il nuovo CSV....")    
        colonne_da_selezionare = ['Order Number', 'Tracking Number', 'Carrier', 'Status']
        df_selezionato = df[colonne_da_selezionare]

        # Crea un buffer di StringIO per salvare il CSV
        output_buffer = StringIO()
        df_selezionato.to_csv(output_buffer, index=False)
        output_buffer.seek(0)  # Torna all'inizio del buffer

        # Salva il nuovo file CSV nella cartella di destinazione
        with sftp.open(output_file_path, 'w') as remote_output_file:
            remote_output_file.write(output_buffer.getvalue())

        print(f"File CSV selezionato salvato con successo in: {output_file_path}")

        # Seleziona solo le colonne richieste 
        print("Attendo 5 minuti per nuovo aggiornamento....") 

    except paramiko.AuthenticationException:
        print("Errore di autenticazione. Verifica il nome utente e la password.")
    except FileNotFoundError as e:
        print(f"Errore: il file o la cartella non esiste: {e}")
    except Exception as e:
        print(f"Errore durante la connessione o l'elaborazione: {e}")
    finally:
        # Chiudi la connessione SFTP
        if transport:
            transport.close()

# Schedula il salvataggio ogni 5 minuti
schedule.every(5).minutes.do(salva_csv)

# Loop per eseguire la schedulazione
while True:
    schedule.run_pending()
    time.sleep(60)  # Attendi un minuto prima di controllare di nuovo

